User-guide: Add automated cleaning feature user document

fmuyassarov opened the pull request:
fixes: https://github.com/metal3-io/metal3-docs/issues/218 and https://github.com/metal3-io/metal3-docs/issues/222

Xenwar commented on docs/user-guide/src/capm3/automated_cleaning.md:
```diff
@@ -0,0 +1,35 @@
+Here are the steps to achieve that:
+
+1. Unset `automatedCleaningMode` in the Metal3MachineTemplate. Then CAPM3 controller unsets it for child Metal3Machines. Although it is unset in the Metal3Machine, BareMetalHosts will get their default automated cleaning mode `metadata`. As we mentioned earlier, CAPM3 controller replicates cleaning mode from Metal3MachineTemplate to Metal3Machine ONLY when it is either `metadata` or `disabled`. A ...
+1. Set `automatedCleaningMode` to `disabled` on the worker Metal3Machine `spec` and to `metadata` on the controlplane Metal3Machine `spec`. Since we don't have any mode set on the Metal3MachineTemplate, Metal3Machines can have different automated cleaning modes set even if they reference the same Metal3MachineTemplate. CAPM3 controller copies cleaning modes from Metal3Machines to their correspond ...
```
```suggestion
1. Set `automatedCleaningMode` to `disabled` on the worker Metal3Machine `spec` and to `metadata` on the controlplane Metal3Machine `spec`. Since we don't have any mode set on the Metal3MachineTemplate, Metal3Machines can have different automated cleaning modes set even if they reference the same Metal3MachineTemplate. CAPM3 controller copies cleaning modes from Metal3Machines to their corresponding BareMetalHosts. As such, we end up with two nodes having different cleaning modes regardless of the fact that they referencing the same Metal3MachineTemplate.
```

Xenwar commented on docs/user-guide/src/capm3/automated_cleaning.md:
```diff
@@ -0,0 +1,35 @@
+
+Here are the steps to achieve that:
+
+1. Unset `automatedCleaningMode` in the Metal3MachineTemplate. Then CAPM3 controller unsets it for child Metal3Machines. Although it is unset in the Metal3Machine, BareMetalHosts will get their default automated cleaning mode `metadata`. As we mentioned earlier, CAPM3 controller replicates cleaning mode from Metal3MachineTemplate to Metal3Machine ONLY when it is either `metadata` or `disabled`. A ...
```
```suggestion
1. Unset `automatedCleaningMode` in the Metal3MachineTemplate. Then CAPM3 controller unsets it for referencing Metal3Machines. Although it is unset in the Metal3Machine, BareMetalHosts will get their default automated cleaning mode `metadata`. As we mentioned earlier, CAPM3 controller replicates cleaning mode from Metal3MachineTemplate to Metal3Machine ONLY when it is either `metadata` or `disabled`. As such, to block synchronization between Metal3MachineTemplate and Metal3Machine, unsetting the cleaning mode in the Metal3MachineTemplate is enough.
```

lentzi90 commented on docs/user-guide/src/bmo/automated_cleaning.md:
```diff
@@ -0,0 +1,32 @@
+
+There are two automated cleaning modes available which can be set via `automatedCleaningMode` field of a BareMetalHost `spec`.
+
+- `metadata` to enable the cleaning
```
I think this deserves an explanation. Why is it `metadata` to enable instead of just `enable`?

fmuyassarov replied:
Added.

lentzi90 commented on docs/user-guide/src/capm3/automated_cleaning.md:
```diff
@@ -0,0 +1,35 @@
+CAPM3 controller ensures to replicate automated cleaning mode to all Metal3Machines from their referenced Metal3MachineTemplate.
+For example, one controlplane and one worker Metal3Machines have `automatedCleaningMode` set to `disabled`, because it is set to `disabled` also in the template that they both are referencing.
+
+**Note**: CAPM3 controller replicates the cleaning mode from Metal3MachineTemplate to Metal3Machine only if `automatedCleaningMode` is not unset (empty). In other words, it synchronizes either `disabled` or `metadata` modes between Metal3MachineTemplate and Metal3Machines.
```
```suggestion
**Note**: CAPM3 controller replicates the cleaning mode from Metal3MachineTemplate to Metal3Machine only if `automatedCleaningMode` is set (not empty). In other words, it synchronizes either `disabled` or `metadata` modes between Metal3MachineTemplate and Metal3Machines.
```

lentzi90 commented on docs/user-guide/src/capm3/automated_cleaning.md:
```diff
@@ -0,0 +1,35 @@
+
+When enabled (`metadata`), automated cleaning kicks off when a node is in the first provisioning and on every deprovisioning.
+There is no defaulting for `automatedCleaningMode` in Metal3MachineTemplate and Metal3Machine. If user doesn't set any mode,
+the field in the `spec` will be omitted. Unsetting `automatedCleaningMode` in the Metal3MachineTemplate will block the synchronization of the cleaning mode between the Metal3MachineTemplate and Metal3Machines.
```
Maybe add here that "this enables the selective operations described below" or "this allows the Metal3Machines to have different values if desired"?
It wasn't clear to me before reading about the selective operations if "blocking the synchronization" was a "configuration error" or something that may be desired in some situations. With this addition I think it would be clearer.

fmuyassarov replied:
good point. Added

lentzi90 reviewed (commented):
Nice job! I added some suggestions.

smoshiur1237 commented on docs/user-guide/src/capm3/automated_cleaning.md:
```diff
@@ -0,0 +1,35 @@
+- `metadata` to enable the cleaning
+- `disabled` to disable the cleaning
+
+When enabled (`metadata`), automated cleaning kicks off when a node is in the first provisioning and on every deprovisioning.
```
```suggestion
When enabled (`metadata`), automated cleaning kicks off when a node is provisioning for the first  and on every time it is deprovisioning.
```

dtantsur commented on docs/user-guide/src/bmo/automated_cleaning.md:
```diff
@@ -0,0 +1,32 @@
+# Automated Cleaning
+
+One of the Ironic's feature exposed to Metal3 Baremetal Operator is [node automated cleaning](https://docs.openstack.org/ironic/latest/admin/cleaning.html#automated-cleaning). When enabled, automated cleaning kicks off while a node is in the first provisioning and on every deprovisioning.
```
s/on every time/every time/.

strictly speaking, passive voice is more appropriate  "node is provisioned"/"deprovisioned"

dtantsur replied:
And should we maybe mention "disk cleaning" at least once?

furkatgofurov7 commented on docs/user-guide/src/capm3/automated_cleaning.md:
```diff
@@ -0,0 +1,35 @@
+# Automated Cleaning
+
+Before reading this page, please see [Baremetal Operator Automated Cleaning](../bmo/automated_cleaning.md) page.
+
+If you are using only Metal3 Baremetal Operator, you can skip this page and refer to Baremetal
+Operator automated cleaning [page](../bmo/automated_cleaning.md) instead.
```
I think these two paragraphs can be merged into one since both are referring/pointing out to same doc and on the same direction

fmuyassarov replied:
Would be nice, but not sure how. Because one is as prerequisite and the letter is for only BMO

fmuyassarov commented:
Thanks folks for reviews. All the comments have been addresses. PTAL once more.

lentzi90 reviewed (approved):
I'm happy with this.
One thing we could add still is an example for the "selective operations" with yaml for Metal3MachineTemplate, Metal3Machines and BareMetalHosts. Then it would be very clear and easy to understand I think.

fmuyassarov commented:
> [APPROVALNOTIFIER] This PR is **NOT APPROVED**
> 
> This pull-request has been approved by: _[lentzi90](https://github.com/metal3-io/metal3-docs/pull/242#pullrequestreview-894813578)_ To complete the [p](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

what do you think about something like this
![references](https://user-images.githubusercontent.com/35802557/156133228-54cec7ef-b83c-4676-b651-58a72ba96b02.png)

lentzi90 commented:
That looks awesome @fmuyassarov!

fmuyassarov commented:
> That looks awesome @fmuyassarov!

Cool. Added it in this patch.

kashifest commented:
/lgtm

fmuyassarov commented:
/cc @dtantsur for approval if looks good :)

fmuyassarov commented:
/hold
I am actually not sure if it is a good idea to add those terminal screenshots as png/svg format, because it is hard to maintain them in the future. If there is a change for the example in the CRD, then there is no programmatic way of updating those images. In general for diagrams I was thinking to use https://plantuml.com/, as for example @Xenwar did in https://github.com/metal3-io/metal3-docs/pull/237. Having declarative definitions of images are much easier to update and maintain. What do you think @lentzi90  ?

lentzi90 commented:
Good point @fmuyassarov and plantuml sounds noce! Can netlify render the diagrams automatically or do we need to do that manually and then upload the result?

fmuyassarov commented:
> Good point @fmuyassarov and plantuml sounds nice! Can netlify render the diagrams automatically or do we need to do that manually and then upload the result?

I think it has to be mdbook processor to render it. Because in our case Netlify just published all the HTML content generated by the mdbook. I will check if mdbook has such a processor, if not we can perhaps upload both result and the script.

fmuyassarov commented:
@lentzi90 I've added the similar diagram as in https://github.com/metal3-io/metal3-docs/pull/242#issuecomment-1055154999 but in plantuml and added image script as well.  You can see the rendered version here https://deploy-preview-242--metal3io.netlify.app/capm3/automated_cleaning.html

lentzi90 commented:
@fmuyassarov great! This looks good to me!

fmuyassarov commented:
/hold cancel
